upf/pfcp.c

```diff
@@ -794,54 +794,50 @@ decode_f_teid (u8 * data, u16 length, void *p)
 {
   pfcp_f_teid_t *v = p;
 
-  if (length < 5)
-    return PFCP_CAUSE_INVALID_LENGTH;
+  if (length < 1)
+   return PFCP_CAUSE_INVALID_LENGTH;
 
   v->flags = get_u8 (data) & 0x0f;
-  v->teid = get_u32 (data);
-  length -= 5;
 
-  if (v->flags & F_TEID_CH)
-    {
-      if (v->flags & (F_TEID_V4 | F_TEID_V6))
-	{
-	  pfcp_debug ("PFCP: F-TEID with invalid flags (CH and v4/v6): %02x.",
-		      v->flags);
-	  return -1;
-	}
-    }
-  else
+  if (!(v->flags & F_TEID_CH))
     {
+      if (length < 4)
+        return PFCP_CAUSE_INVALID_LENGTH;
+      v->teid = get_u32 (data);
+      length -= 4;
+
       if (v->flags & F_TEID_CHID)
-	{
-	  pfcp_debug
-	    ("PFCP: F-TEID with invalid flags (CHID without CH): %02x.",
-	     v->flags);
-	  return -1;
-	}
+        {
+          pfcp_debug
+            ("PFCP: F-TEID with invalid flags (CHID without CH): %02x.",
+             v->flags);
+          return -1;
+        }
+
       if (!(v->flags & (F_TEID_V4 | F_TEID_V6)))
-	{
-	  pfcp_debug ("PFCP: F-TEID without v4/v6 address: %02x.", v->flags);
-	  return -1;
-	}
-    }
+        {
+          pfcp_debug ("PFCP: F-TEID without v4/v6 address: %02x.", v->flags);
+          return -1;
+        }
 
-  if (v->flags & F_TEID_V4)
-    {
-      if (length < 4)
-	return PFCP_CAUSE_INVALID_LENGTH;
+        if (v->flags & F_TEID_V4)
+          {
+            if (length < 4)
+              return PFCP_CAUSE_INVALID_LENGTH;
 
-      get_ip4 (v->ip4, data);
-      length -= 4;
-    }
+            get_ip4 (v->ip4, data);
+            length -= 4;
+          }
 
-  if (v->flags & F_TEID_V6)
-    {
-      if (length < 16)
-	return PFCP_CAUSE_INVALID_LENGTH;
+        if (v->flags & F_TEID_V6)
+          {
+            if (length < 16)
+              return PFCP_CAUSE_INVALID_LENGTH;
+
+            get_ip6 (v->ip6, data);
+            length -= 16;
+          }
 
-      get_ip6 (v->ip6, data);
-      length -= 16;
     }
 
   if (v->flags & F_TEID_CHID)
```

Comment:
hmm, am I getting this wrong?

– Bit 1 – V4: If this bit is set to "1" and the CH bit is not set, then the IPv4 address field shall be present, otherwise the IPv4 address field shall not be present.

– Bit 2 – V6: If this bit is set to "1" and the CH bit is not set, then the IPv6 address field shall be present, otherwise the IPv6 address field shall not be present.

In old version, if CH is set flags v4/v6 can't be present in IE_FTEID. However, in this case if CH is set one/both of version flags can be present. From my point of view the only check we can plant here looks like this: if CH is set and one/both version flags are set, there should be no IP(4/6) field present in a IE_FTEID. I will put this check.

Reply:
If CH flag is set, the v4/v6 field can not be present, however the flags can be set. In fact, independent of who CH ist set, at least on the v4/v6 flags need to be set.

TS 29.244, V16.4.0 has this:

> At least one of the V4 and V6 flags shall be set to "1", and both may be set to "1" for both scenarios:
> -	when the CP function is providing F-TEID, i.e. both IPv4 address field and IPv6 address field may be present; or
> -	when the UP function is requested to allocate the F-TEID, i.e. when CHOOSE bit is set to "1", and the IPv4 address and IPv6 address fields are not present.